cirq/ops/linear_combinations.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Dict, Mapping, Optional, Tuple
+
+import numpy as np
+
+from cirq import protocols, value
+from cirq.ops import raw_types
+
+
+class LinearCombinationOfGates(value.LinearDict[raw_types.Gate]):
+    """Represents linear operator defined by a linear combination of gates.
+
+    Note that the operator may not be unitary or even normal. Examples of
+    operators that may be represented as LinearCombinationOfGates:
+     * Hamiltonians,
+     * measurement operators,
+     * Kraus operators,
+     * creation and annihilation operators.
+    """
+    def __init__(self, terms: Mapping[raw_types.Gate, value.Scalar]) -> None:
+        """Initializes linear combination from a collection of terms.
+
+        Args:
+            terms: Mapping of gates to coefficients in the linear combination
+                being initialized.
+        """
+        super().__init__(terms, validator=self._is_compatible)
+
+    def num_qubits(self) -> Optional[int]:
+        """Returns number of qubits in the domain if known, None if unknown."""
+        if not self:
+            return None
+        any_gate = next(iter(self))
+        return any_gate.num_qubits()
+
+    def _is_compatible(self, gate: raw_types.Gate) -> bool:
+        return (self.num_qubits() is None or
+                self.num_qubits() == gate.num_qubits())
+
+    def matrix(self) -> np.ndarray:
+        num_qubits = self.num_qubits()
+        if num_qubits is None:
+            raise ValueError('Unknown number of qubits')
+        num_dim = 2 ** num_qubits
+        result = np.zeros((num_dim, num_dim), dtype=np.complex128)
+        for gate, coefficient in self.items():
+            result += protocols.unitary(gate) * coefficient
```

Comment:
Some gates aren't unitary. Is this class only for unitary gates? Maybe that should be in the name of the class, or in the name of this method e.g. `sum_of_unitaries` or `total_matrix` or `combined_matrix` or `combined_matrix_from_unitaries`...

@dabacon Is there something reasonable we can do when the user passes in a general channel?

Reply:
I think for linear combinations with non-unitary terms, we should just raise an exception. Present code does this implicitly.

Reply:
We could include `has_unitary` in our validator.

Reply:
I think it should be legal to create linear combinations of gates without unitaries. However, a user of such a linear combination accepts that `matrix()` raises an exception. Added comment to make this contract explicit.